httpapi/httpapi.go

```diff
@@ -23,16 +41,39 @@ func NewServeMux(nodes []*node.HydraNode) *http.ServeMux {
 				Addrs: n.Host.Addrs(),
 			})
 		}
-	})
+	}
+}
 
-	return mux
+// "/records/fetch" Receive a record and fetch it from the network, if available
+func recordFetchHandler(nodes []*node.HydraNode) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// TODO needs functionality to be implemented in libp2p
+		// See: https://discuss.libp2p.io/t/does-a-findproviders-replicate-the-provider-records-to-the-node-issuing-the-query/452
+		httpNotImplemented := 501
+		w.WriteHeader(httpNotImplemented)
+	}
 }
 
-// ListenAndServe instructs a Hydra HTTP API server to listen and serve on the passed address
-func ListenAndServe(nodes []*node.HydraNode, addr string) error {
-	listener, err := net.Listen("tcp", addr)
-	if err != nil {
-		return err
+// "/records/list" Receive a record and fetch it from the network, if available
+func recordListHandler(nodes []*node.HydraNode, datastore ds.Batching) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// TODO Improve this handler once ProvideManager gets exposed
+		// https://discuss.libp2p.io/t/list-provider-records/450
+		// for now, enumerate the Provider Records in the datastore
+
+		providersKeyPrefix := "/providers/" // https://github.com/libp2p/go-libp2p-kad-dht/blob/master/providers/providers.go#L76
+		results, err := datastore.Query(dsq.Query{
```

Comment:
I know this is weird, but we know they share the same datastore...

```suggestion
		results, err := nodes[0].Datastore.Query(dsq.Query{
```

Do you think we need a:

```go
type Hydra struct {
	Heads		[]*head.Head // or maybe `Sybils`?
	Datastore	datastore.Datastore
	RoutingTable	*kbucket.RoutingTable
}
```

? Where `head.Head` is basically what `HydraNode` is now:

```go
type Head struct {
	Host         host.Host
	Routing      routing.Routing
	Bootstrapped bool
}
```

...and we'd pass our `Hydra` instance to `NewServeMux`.

Maybe one for a separate PR?

Reply:
I think that is a premature refactor. Right now, the cost is just one repeated pointer.